### ISSUE TITLE
Out of Resources Error with plenty of Memory

### ISSUE THREAD

**Issue by cdeterman** [viennacl/viennacl-dev#222]
I am trying to troubleshoot an error.  I try to monitor the GPU memory usage quite closely to make sure I don't have any memory leaks that take up all the memory.  In my case, I am using a nvidia GPU so I can use `watch nvidia-smi` command.

That said, I have run in to a situation which results in the following error when I try to instantiate another `viennacl::vector` of any size.

>ViennaCL: FATAL ERROR: CL_OUT_OF_RESOURCES 
 ViennaCL tried to launch a compute kernel, but the device does not provide enough resources. Try changing the global and local work item sizes.

Now, in the past I would look at the `nvidia-smi` output and it would indeed have almost all the RAM taken up.  But in this case it is barely gets above 470MB when ~4GB are available.  How could this be?  Is there any other reason that error would be thrown?  Any insight in to troubleshooting this would be appreciated.

**Comment by cdeterman** [comment]
It actually appears to be coming from the `fast_copy` call I have.   Any thoughts on troubleshooting a call like this excerpt?

```c++
Eigen::Matrix<T, Eigen::Dynamic, 1> Am;

int K = Am.size();
           
viennacl::vector_base<T> A = viennacl::vector_base<T>(K); 

viennacl::fast_copy(Am.data(), Am.data() + Am.size(), A.begin());
```

**Comment by karlrupp** [comment]
Most likely you get the error because of some previous memory corruption. Any chance that you can run the code through valgrind?

Also, is the template type T an integer or a floating point type?

**Comment by cdeterman** [comment]
The `T` is a floating point type.  I can give valgrind a shot.  Hopefully something will stick out.  I always have a rough time interpreting the output.

**Comment by cdeterman** [comment]
@karlrupp I found two pointers that were leaking that I have resolved but unfortunately hasn't fixed the problem.  I know you are always busy too but would you be willing to look at the valgrind output?  It is quite long though.

The really frustrating part is that it is only happening in a `for` loop within a `while` statement.   It completes the for loop but when the next `while` iteration happens the `for` loop crashes on that `fast_copy` call but it will still make other `vector` objects not initialized like that.  It is extremely confusing.  Thanks for any help.

**Comment by cdeterman** [comment]
Further details, it appears that once this error comes up any calls to any `copy` or `fast_copy` fail for both `vector` and `matrix` objects.  And yet, if I initialize with `zero_matrix` or `zero_vector` the initialization still works without a problem.  Not sure if this helps at all but at least it helps narrow it down somewhat.

**Comment by cdeterman** [comment]
Ugh... I finally sorted it out.  Nothing really reported anything useful debugging wise so I resorted to the old fashion method of printing comments and commenting out different parts of the program.  This finally led me to one of my custom OpenCL kernels I was using.  I was accidentally indexing outside of the vector space with the kernel at larger sizes.  Fixing that now allows the program to function without the crashing.  Thanks for suggesting the memory corruption as it led me down this path.
